framework/src/meshgenerators/CutMeshByPlaneGenerator.C

```diff
@@ -0,0 +1,458 @@
+  if (tet4_nodes_outside_plane.size() == 0)
+  {
+    if (tet4_nodes_on_plane.size() == 3)
+    {
+      // Record the element for future cross section boundary assignment
+      elems_Tet4.push_back(mesh.elem_ptr(elem_id));
+    }
+  }
+  // Remove the element if all the nodes are outside the plane
+  else if (tet4_nodes_inside_plane.size() == 0)
+  {
+    mesh.elem_ptr(elem_id)->subdomain_id() = block_id_to_remove;
+    if (tet4_nodes_on_plane.size() == 3)
+    {
+      // I think the neighboring element will be handled,
+      // So we do not need to assign the cross section boundary here
+      // Need recheck
+    }
+  }
+  // As we have nodes on both sides, five different scenarios are possible
+  else
+  {
+    new_elements_created = true;
+    if (tet4_nodes_inside_plane.size() == 1 && tet4_nodes_outside_plane.size() == 3)
+    {
+      std::vector<Node *> new_plane_nodes;
+      // A smaller TET4 element is created, this solution is unique
+      for (const auto & tet4_node_outside_plane : tet4_nodes_outside_plane)
+      {
+        new_plane_nodes.push_back(nonDuplicateNodeCreator(
+            mesh,
+            new_on_plane_nodes,
+            pointPairPlaneInterception(
+                *tet4_node_outside_plane, *tet4_nodes_inside_plane[0], plane_point, plane_normal)));
+      }
+      elems_Tet4.push_back(mesh.add_elem(new Tet4));
+      elems_Tet4.back()->set_node(0) = tet4_nodes_inside_plane[0];
+      elems_Tet4.back()->set_node(1) = new_plane_nodes[0];
+      elems_Tet4.back()->set_node(2) = new_plane_nodes[1];
+      elems_Tet4.back()->set_node(3) = new_plane_nodes[2];
+      elems_Tet4.back()->subdomain_id() = mesh.elem_ptr(elem_id)->subdomain_id();
+    }
+    else if (tet4_nodes_inside_plane.size() == 2 && tet4_nodes_outside_plane.size() == 2)
+    {
+      std::vector<Node *> new_plane_nodes;
+      // 3 smaller TET3 elements are created
+      for (const auto & tet4_node_outside_plane : tet4_nodes_outside_plane)
+      {
+        for (const auto & tet4_node_inside_plane : tet4_nodes_inside_plane)
+        {
+          new_plane_nodes.push_back(nonDuplicateNodeCreator(
+              mesh,
+              new_on_plane_nodes,
+              pointPairPlaneInterception(
+                  *tet4_node_outside_plane, *tet4_node_inside_plane, plane_point, plane_normal)));
+        }
+      }
+      std::vector<Node *> new_elems_nodes = {tet4_nodes_inside_plane[1],
+                                             new_plane_nodes[3],
+                                             new_plane_nodes[1],
+                                             tet4_nodes_inside_plane[0],
+                                             new_plane_nodes[2],
+                                             new_plane_nodes[0]};
+      std::vector<std::vector<unsigned int>> rotated_tet_face_indices;
+      auto optimized_node_list =
+          MooseMeshUtils::prismNodeOptimizer(new_elems_nodes, rotated_tet_face_indices);
```

Comment:
how do you know it was a prism

Reply:
<img width="310" alt="image" src="https://github.com/idaholab/moose/assets/57006315/faa64afe-5ccd-45d3-ad2d-ff66d31bb8dd">
It's this case.